include/mpscringbuffer.hpp

```diff
@@ -0,0 +1,52 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <thread>
+
+namespace ddprof {
+//  mimic: std::hardware_destructive_interference_size, C++17
+constexpr std::size_t hardware_destructive_interference_size = 128;
+
+struct MPSCRingBufferMetaDataPage {
+  alignas(hardware_destructive_interference_size) uint64_t writer_pos;
+  alignas(hardware_destructive_interference_size) uint64_t reader_pos;
+  alignas(hardware_destructive_interference_size) std::atomic_flag spinlock;
+};
+
+inline void lock(std::atomic_flag *lock) {
+  static constexpr uint32_t k_max_active_spin = 4000;
+  static constexpr std::chrono::nanoseconds k_yield_sleep =
+      std::chrono::microseconds(500);
+
+  uint32_t spincount = 0;
+
+  for (;;) {
+    if (!lock->test_and_set(std::memory_order_acquire)) {
+      break;
+    }
+    do {
+      if (spincount < k_max_active_spin) {
+        ++spincount;
+#ifdef __x86_64__
+        asm volatile("pause");
+#else
+        asm volatile("yield");
+#endif
+      } else {
+        // If active spin fails, yield
+        std::this_thread::sleep_for(k_yield_sleep);
+      }
+      // Wait for lock to be released without generating cache misses
+    } while (lock->test(std::memory_order_relaxed));
+  }
```

Comment:
My understanding is that the order of arrival of the threads and 500 microseconds might allow the threads not to go back to spinning. Though perhaps in some apps all threads arrive at the same time.

Reply:
This type of fast-path/slow-path spinlock is surprisingly durable.  I'd structure a test which sits on a tight loop over the lock in a pool of threads, doubling the number of threads every iteration from 1x to 1024x the number of CPUs and observe the scaling properties of the lock itself after a token amount of on-CPU work.  I'd expect
* Median wait time per thread to remain linear up to a big multiple of the number of cores.
* After breaking linear scaling, we get quadratic rather than exponential

Why do I think this?  Jiffies are big and CPUs are fast.  All bets are off the table if the work loop has a context switch though, in that case I'm with Ivo.

But again, I think the point is to look at the scaling properties rather than a given a priori overhead value for a fixed number of threads.

I do think that ceiling should be evaluated though.  We're adding a spinlock to a library which is trying to infect its children, and some processes are more, uh, prodigious breeders than others.  :)

Reply:
IMHO, this is pretty idiomatic.
1. Use of pause/yield on x86/arm to tell the processor that this is a spin-wait loop
2. Use of OS sched yield to sleep thread after fixed number of iterations in the fast path, back to lock

One of the problems with the classic spinlock is, as Ivo notes, suppose that the spinlock is the only thing on CPU, with a bunch of on- and off-cpu threads in contention.  Draining the spinlock will depend on the active thread being able to progress, and whenever it has to yield to the scheduler it might have to round-robin behind all of the waiting threads as they clash on the lock.

Reply:
Ahhhh you're right about `spincount`, I completely missed that part and read it as the loop starting as-if from scratch 😅